### PR TITLE
[ENGINE] Adding more Spans

### PR DESCRIPTION
**Pull request by Alfred-Mountfield**
## 🌟 What is the purpose of this PR?

Expands upon the work done in #228 by adding more spans around key parts of code execution. 
This greatly improves the timings produced by TeXRay, and adds additional information to general logging.

It also cleans up some of the logging in general, where we have messages that are far too large to be printed.

## 🔍 What does this change?
- Overrides the Debug impl for the `SuccessMessage` in the JsPy init package, as we shouldn't be printing the full Agent JSON, that can be far too large
- Removes a trace log in the Javascript runner where we log every message (including init messages like above) which can be far too large and pollutes the logs.
- Adds Spans for the following
  - Running Init Packages 
  - Preparing for Context Packages
  - Running Context Packages
    - A span for each package
      - A span for the run method
      - A span for the write method
  - Running State Packages
    - A span for each package 
  - Running Output Packages
    - A span for each package 

To do this it adds a `get_span` method on most Packages so they can provide the span they want wrapped around their execution (this is necessary as we have trait objects for Packages so we can't set it manually)

### Does this require a change to the docs?

Nope

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201481007343159/1201734259001872/f) (_internal_)

## 🛡 What tests cover this?

- This shouldn't affect functionalities, so the normal integration tests passing is sufficient

## ❓ How to test this?

1. Checkout the branch / view the deployment
2. Run a simulation with `RUST_LOG=info` and look for spans on events happening in the situations mentioned above, then run with `RUST_LOG=trace` and note the additional span information that's provided
3. Optionally run with `--features "texray"` for both cases above and look at the output in the log folder.

### Logs in Debug
image

### Logs in Trace
image

### TeXRay in Info
image

### TeXRay in Debug
image

### TeXRay in Trace
image

### CODE REVIEW

File: packages/engine/src/simulation/package/context/packages/api_requests/mod.rs [hunk 1]
```diff
@@ -196,3 +190,32 @@ pub fn custom_message_handlers_from_properties(
         })
         .transpose()
 }
+
+async fn build_api_response_maps(
+    snapshot: &StateSnapshot,
+    handlers: &[String],
+) -> Result<Vec<ApiResponseMap>> {
+    tracing::warn!("BUILDIN API RESPONSE");
```
Comment (TimDiekmann) [inline packages/engine/src/simulation/package/context/packages/api_requests/mod.rs:198]: What is this warning about? Looks like a println-style debug output.

File: packages/engine/src/simulation/package/context/packages/neighbors/mod.rs [hunk 2]
```diff
@@ -162,4 +169,8 @@ impl Package for Neighbors {
 
         Ok(vec![(field_key, Arc::new(neighbors_builder.finish()))])
     }
+
+    fn get_span(&self) -> Span {
```
Comment (TimDiekmann) [inline packages/engine/src/simulation/package/context/packages/neighbors/mod.rs:173]: In most cases, getters in Rust omit `get_`.
```suggestion
    fn span(&self) -> Span {
```

File: packages/engine/src/simulation/package/output/mod.rs [hunk 3]
```diff
@@ -60,4 +61,6 @@ pub trait PackageCreator: GetWorkerExpStartMsg + Sync + Send {
 #[async_trait]
 pub trait Package: MaybeCpuBound + GetWorkerSimStartMsg + Send + Sync {
     async fn run(&mut self, state: Arc<State>, context: Arc<Context>) -> Result<Output>;
+
+    fn get_span(&self) -> Span;
```
Comment (TimDiekmann) [inline packages/engine/src/simulation/package/output/mod.rs:65]:
```suggestion
    fn span(&self) -> Span;
```

File: packages/engine/src/simulation/package/output/packages/analysis/mod.rs [hunk 4]
```diff
@@ -87,6 +88,10 @@ impl Package for Analysis {
             self.analyzer.get_latest_output_set(),
         ))
     }
+
+    fn get_span(&self) -> Span {
```
Comment (TimDiekmann) [inline packages/engine/src/simulation/package/output/packages/analysis/mod.rs:92]:
```suggestion
    fn span(&self) -> Span {
```

File: packages/engine/src/simulation/package/output/packages/json_state/mod.rs [hunk 5]
```diff
@@ -107,6 +107,10 @@ impl Package for JsonState {
             inner: agent_states,
         }))
     }
+
+    fn get_span(&self) -> Span {
```
Comment (TimDiekmann) [inline packages/engine/src/simulation/package/output/packages/json_state/mod.rs:111]:
```suggestion
    fn span(&self) -> Span {
```

File: packages/engine/src/simulation/package/context/mod.rs [hunk 6]
```diff
@@ -36,6 +37,8 @@ pub trait Package: MaybeCpuBound + GetWorkerSimStartMsg + Send + Sync {
         num_agents: usize,
         context_schema: &ContextSchema,
     ) -> Result<Vec<(FieldKey, Arc<dyn arrow::array::Array>)>>;
+
+    fn get_span(&self) -> Span;
```
Comment (TimDiekmann) [inline packages/engine/src/simulation/package/context/mod.rs:41]:
```suggestion
    fn span(&self) -> Span;
```

File: packages/engine/src/simulation/package/context/packages/api_requests/mod.rs [hunk 7]
```diff
@@ -88,33 +89,21 @@ impl Package for ApiRequests {
...
-                        let messages = handlers::gather_requests(&reader, messages)?;
-                        futs.push(handlers::run_custom_message_handler(handler, messages))
-                    }
-                    Ok(())
-                })?;
-            }
+        // We want to pass the span for the package to the writer, so that the write() call isn't
+        // nested under the run span
+        let pkg_span = Span::current();
+        let run_span = tracing::trace_span!("run"); // store an un-entered span for the async
```
Comment (TimDiekmann) [inline packages/engine/src/simulation/package/context/packages/api_requests/mod.rs:95]:
```suggestion
        // store an un-entered span for the async
        let run_span = tracing::trace_span!("run");
```

File: packages/engine/src/simulation/package/run.rs [hunk 8]
```diff
@@ -168,14 +168,21 @@ impl StepPackages {
 
             let cpu_bound = package.cpu_bound();
             futs.push(if cpu_bound {
+                let current_span = Span::current();
                 tokio::task::spawn_blocking(move || {
-                    let res = block_on(package.run(state, snapshot_clone).in_current_span());
+                    let package_span = {
+                        // We want to create the package span within the scope of the current one
+                        let _entered = current_span.entered();
+                        package.get_span()
```
Comment (TimDiekmann) [inline packages/engine/src/simulation/package/run.rs:176]:
```suggestion
                        package.span()
```

File: packages/engine/src/simulation/package/run.rs [hunk 9]
```diff
@@ -168,14 +168,21 @@ impl StepPackages {
...
+                    };
+                    let res = block_on(package.run(state, snapshot_clone).instrument(package_span));
                     (package, res)
                 })
             } else {
+                let span = package.get_span();
```
Comment (TimDiekmann) [inline packages/engine/src/simulation/package/run.rs:182]:
```suggestion
                let span = package.span();
```

File: packages/engine/src/simulation/package/run.rs [hunk 10]
```diff
@@ -241,7 +248,8 @@ impl StepPackages {
         // Traits are tricky anyway for working with iterators
         // Will instead use state.upgrade() and exstate.downgrade() and respectively for context
         for pkg in self.state.iter_mut() {
-            pkg.run(&mut state, context).await?;
+            let span = pkg.get_span();
```
Comment (TimDiekmann) [inline packages/engine/src/simulation/package/run.rs:251]:
```suggestion
            let span = pkg.span();
```

File: packages/engine/src/simulation/package/run.rs [hunk 11]
```diff
@@ -263,15 +271,25 @@ impl StepPackages {
 
             let cpu_bound = pkg.cpu_bound();
             futs.push(if cpu_bound {
+                let current_span = Span::current();
                 tokio::task::spawn_blocking(move || {
-                    let res = block_on(pkg.run(state, context));
+                    let package_span = {
+                        // We want to create the package span within the scope of the current one
+                        let _entered = current_span.entered();
+                        pkg.get_span()
```
Comment (TimDiekmann) [inline packages/engine/src/simulation/package/run.rs:279]:
```suggestion
                        pkg.span()
```

File: packages/engine/src/simulation/package/run.rs [hunk 12]
```diff
@@ -263,15 +271,25 @@ impl StepPackages {
...
+                    };
+                    let res = block_on(pkg.run(state, context).instrument(package_span));
                     (pkg, res)
                 })
             } else {
-                tokio::task::spawn(async {
-                    let res = pkg.run(state, context).await;
-                    (pkg, res)
-                })
+                let span = pkg.get_span();
```
Comment (TimDiekmann) [inline packages/engine/src/simulation/package/run.rs:285]:
```suggestion
                let span = pkg.span();
```

File: packages/engine/src/simulation/package/state/mod.rs [hunk 13]
```diff
@@ -23,6 +24,8 @@ use crate::{
 #[async_trait]
 pub trait Package: GetWorkerSimStartMsg + Send + Sync {
     async fn run(&mut self, state: &mut StateMut, context: &Context) -> Result<()>;
+
+    fn get_span(&self) -> Span;
```
Comment (TimDiekmann) [inline packages/engine/src/simulation/package/state/mod.rs:28]:
```suggestion
    fn span(&self) -> Span;
```

File: packages/engine/src/simulation/package/state/packages/behavior_execution/mod.rs [hunk 14]
```diff
@@ -234,4 +237,8 @@ impl Package for BehaviorExecution {
         tracing::trace!("BehaviorExecution task finished: {:?}", &msg);
         Ok(())
     }
+
+    fn get_span(&self) -> Span {
```
Comment (TimDiekmann) [inline packages/engine/src/simulation/package/state/packages/behavior_execution/mod.rs:241]:
```suggestion
    fn span(&self) -> Span {
```

File: packages/engine/src/simulation/package/state/packages/topology/mod.rs [hunk 15]
```diff
@@ -89,4 +89,8 @@ impl Package for Topology {
         }
         Ok(())
     }
+
+    fn get_span(&self) -> Span {
```
Comment (TimDiekmann) [inline packages/engine/src/simulation/package/state/packages/topology/mod.rs:93]:
```suggestion
    fn span(&self) -> Span {
```

**Review by TimDiekmann — CHANGES_REQUESTED** [review]
Adding the spans appears to be a bit verbose, but the outputs are definitely worth it!

Looks good to me, only "2" changes I'd suggest.

File: packages/engine/src/simulation/package/context/packages/api_requests/mod.rs (hunk 1 above)
Comment (Alfred-Mountfield) [reply packages/engine/src/simulation/package/context/packages/api_requests/mod.rs:198]: I sloppily left it in from some debugging. Darn. Thanks for spotting.
```suggestion
```

**Review by TimDiekmann — APPROVED** [review]

File: packages/engine/src/simulation/package/context/packages/api_requests/mod.rs (hunk 7 above)
Comment (Alfred-Mountfield) [reply packages/engine/src/simulation/package/context/packages/api_requests/mod.rs:95]: Have chatted about this offline, differing opinion about inline comments

**Pull request merged**